import sys
import random
from PyQt6.QtWidgets import QApplication, QMainWindow
from PyQt6.QtCore import QTimer
from PyQt6.uic import loadUi

class NimGame(QMainWindow):
    def __init__(self):
        super().__init__()
        loadUi('task4.ui', self)
        self.setWindowTitle("–ò–≥—Ä–∞ '–ü—Å–µ–≤–¥–æ–Ω–∏–º'")
        self.pushButton_start.clicked.connect(self.start_game)
        self.pushButton_take.clicked.connect(self.player_move)
        self.pushButton_take.setEnabled(False)
        self.spinBox.setMaximum(3)
        self.spinBox.setMinimum(1)
        self.stones = 0

    def start_game(self):
        try:
            n = int(self.lineEdit.text())
            if n <= 0:
                raise ValueError
            self.stones = n
            self.label_stones.setText(f"–û—Å—Ç–∞–ª–æ—Å—å: {n}")
            self.label_status.setText("–í–∞—à —Ö–æ–¥!")
            self.pushButton_take.setEnabled(True)
            self.lineEdit.setEnabled(False)
            self.pushButton_start.setEnabled(False)
            self.spinBox.setValue(1)
        except:
            self.label_status.setText("–í–≤–µ–¥–∏—Ç–µ –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ–µ —á–∏—Å–ª–æ –∫–∞–º–Ω–µ–π!")

    def player_move(self):
        take = self.spinBox.value()
        if take > self.stones:
            self.label_status.setText("–ù–µ–ª—å–∑—è –≤–∑—è—Ç—å –±–æ–ª—å—à–µ, —á–µ–º –µ—Å—Ç—å!")
            return
        self.stones -= take
        self.label_stones.setText(f"–û—Å—Ç–∞–ª–æ—Å—å: {self.stones}")
        if self.stones == 0:
            self.label_status.setText("üéâ –í—ã –≤—ã–∏–≥—Ä–∞–ª–∏!")
            self.end_game()
            return
        self.pushButton_take.setEnabled(False)
        self.label_status.setText("–•–æ–¥ –∫–æ–º–ø—å—é—Ç–µ—Ä–∞...")
        QTimer.singleShot(1000, self.computer_move)

    def computer_move(self):
        if self.stones <= 3:
            take = self.stones
        else:
            take = (self.stones - 1) % 4
            if take == 0:
                take = 1
        self.stones -= take
        self.label_stones.setText(f"–û—Å—Ç–∞–ª–æ—Å—å: {self.stones}")
        if self.stones == 0:
            self.label_status.setText("üíÄ –ö–æ–º–ø—å—é—Ç–µ—Ä –≤—ã–∏–≥—Ä–∞–ª!")
            self.end_game()
            return
        self.pushButton_take.setEnabled(True)
        self.label_status.setText("–í–∞—à —Ö–æ–¥!")

    def end_game(self):
        self.pushButton_take.setEnabled(False)
        self.lineEdit.setEnabled(True)
        self.pushButton_start.setEnabled(True)

if __name__ == '__main__':
    app = QApplication(sys.argv)
    window = NimGame()
    window.show()
    sys.exit(app.exec())